Implement optimization bisect

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49031 Implement optimization bisect**

Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

tugsbayasgalan committed:
Implement optimization bisect

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.h:
```diff
@@ -0,0 +1,60 @@
+// Bisecting is enabled on a per file basis (hence per pass). For example, in
+// `constant_propagation.cpp`, `PYTORCH_OPT_LIMIT` should be
+// set to `constant_propagation.cpp=<opt_limt>` or, simply, to
+// `constant_propagation.cpp=<opt_limit>` where <opt_limit> is the number of
```
Nit: did you mean `constant_propagation=<opt_limit>` here (without `.cpp`)?

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.h:
```diff
@@ -0,0 +1,60 @@
+
+#define JIT_BISECT(...) (is_bisect_enabled(__FILE__));
+
+#define JIT_BISECT_LOG(MSG, ...)                           \
```
What's the difference between `JIT_BISECT_LOG` and `GRAPH_DUMP` and friends? If possible, we should avoid duplicating functionality.

tugsbayasgalan replied:
Yeah I think we can use GRAPH_DUMP but I think user has to manually pass in the bisect related information to the log in that case.

ZolotukhinM commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -336,6 +338,13 @@ struct ConstantPropagator {
   void ConstantPropagation(Node* n) {
+    auto allowed = JIT_BISECT();
+    if (!allowed) {
+      JIT_BISECT_LOG(
```
This will probably spam the output after the optimization has reached its limit (i.e. we will keep trying to run the optimization no matter what, but with opt limit after some point each run would simply print this message and return). However, it might useful to detect the **last** invocation of the optimization - would that be possible to do somehow?

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+static std::unordered_map<std::string, int64_t> parseJITBisectOption(
+    const char* option) {
+  std::stringstream in_ss;
+  in_ss << "function:";
```
What is this "function:"? I didn't see it mentioned in the comment.

tugsbayasgalan replied:
oh sorry this was from jit_log.cpp and i forgot to delete

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+    }
+    auto index_at = line.find_last_of('=');
+    auto pass_name = line.substr(0, index_at);
+    pass_name = c10::detail::ExcludeFileExtension(pass_name);
```
Since we're adding a `ExcludeFileExtension` function, it would make sense to also update this place:
https://github.com/pytorch/pytorch/blob/fb755ad33ef84844e4f4c2d285b71737316814f8/torch/csrc/jit/jit_log.cpp#L45-L48

It also might be a good idea to extract it to a separate PR, but up to you.

tugsbayasgalan replied:
sounds good! will do

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.h:
```diff
@@ -0,0 +1,60 @@
+#include <unordered_map>
+
+// `TorchScript` offers a simple bisection utility function
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
```
Nit: since it only affect JIT passes, it might makes sense to include `JIT` into the env var name.

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.h:
```diff
@@ -0,0 +1,60 @@
+
+TORCH_API bool is_bisect_enabled(const char* pass_name);
+
+TORCH_API std::ostream& operator<<(std::ostream& out, int64_t level);
```
Why do we need to redefine this operator for a standard type? Also, it doesn't seem to be implemented anywhere.

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
+  // TODO bad code
```
😃

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+namespace torch {
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
```
Nit: since everywhere else we're using `int64_t` it might make sense to return it here too.

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+#include <torch/csrc/jit/frontend/error_report.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_bisect.h>
+#include <torch/csrc/jit/serialization/python_print.h>
```
What do we need this include for?

tugsbayasgalan replied:
good catch! will remove them

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+
+  auto opt_limit_it = passes_to_opt_limits.find(pass);
+  if (opt_limit_it == passes_to_opt_limits.end()) {
+    return false;
```
Shouldn't we return true (don't limit the optimization) if it was not listed in `PYTORCH_OPT_LIMIT`?

tugsbayasgalan replied:
Ah yep that is true! will fix

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.cpp:
```diff
@@ -0,0 +1,121 @@
+  static const char* opt_limit = std::getenv("PYTORCH_OPT_LIMIT");
+  // if nothing is provided, let's allow everything
+  if (!opt_limit) {
+    return true;
```
Nit: I find the name of the function a bit confusing: it says "bisect is enabled", but when the `PYTORCH_OPT_LIMIT ` flag is not set, then we return `true`.

ZolotukhinM commented on torch/csrc/jit/jit_opt_bisect.h:
```diff
@@ -0,0 +1,60 @@
+#include <string>
+#include <unordered_map>
+
+// `TorchScript` offers a simple bisection utility function
```
We probably also need to add a description to `OVERVIEW.md`: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-logging

ZolotukhinM commented on torch/csrc/jit/passes/dead_code_elimination.cpp:
```diff
@@ -259,6 +260,14 @@ class DeadCodeEliminator {
+    auto allowed = JIT_BISECT();
+    if (!allowed) {
+      JIT_BISECT_LOG(
+        std::string("Stop deleting because of opt limitation"),
```
Nit: generally it might be a good idea to "standardize" such messages across passes.

ZolotukhinM reviewed (commented):
Very nice! I'm super excited to see it's landed and start using it :) I left some comments inline, mostly minor. Once we resolve them, it'll be good to go!

Also, cc @Krovatkin who implemented `jit_log` and might provide some valuable feedback.

ZolotukhinM commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -336,6 +338,13 @@ struct ConstantPropagator {
   }
 
   void ConstantPropagation(Node* n) {
+    auto allowed = JIT_BISECT();
```
Nit: a random thought, I've just realized that while this API will probably mostly be used for the bisection, the macro itself doesn't do any bisection: it simply enables/disables the optimization. Maybe we should pick a different name that better reflects what it does?

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan commented:
I will submit a seperate PR to use ExcludeFileExtension on jit_log since this PR is not related to jit_log

tugsbayasgalan committed:
Update on "[WIP] Implement optimization bisect"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/jit_opt_limit.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
+  // TODO bad code
```
Please address this todo :) Or expand on what exactly needs to be done.

ZolotukhinM commented on torch/csrc/jit/jit_opt_limit.h:
```diff
@@ -0,0 +1,46 @@
+#include <unordered_map>
+
+// `TorchScript` offers a simple optimization limit checker
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
```
Nit: `PYTORCH_JIT_OPT_LIMIT`

ZolotukhinM commented on torch/csrc/jit/jit_opt_limit.h:
```diff
@@ -0,0 +1,46 @@
+// `TorchScript` offers a simple optimization limit checker
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
+// The purpose is to limit how many optimization you can make per pass.
+// This is useful for debugging performance passes.
```
Nit: not only performance, any passes :)

ZolotukhinM commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <torch/csrc/jit/runtime/operator.h>
 #include <torch/csrc/jit/runtime/vararg_functions.h>
 #include <torch/csrc/utils/memory.h>
+#include <string>
```
Nit: Unnecessary change.

ZolotukhinM commented on torch/csrc/jit/jit_opt_limit.h:
```diff
@@ -0,0 +1,46 @@
+namespace torch {
+namespace jit {
+
+struct Node;
+struct Graph;
```
Nit: These forward decls are not used.

ZolotukhinM commented on torch/csrc/jit/jit_opt_limit.h:
```diff
@@ -0,0 +1,46 @@
+#pragma once
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
```
Nit: unused?

ZolotukhinM reviewed (approved):
Awesome work, thanks! I only have some nitpicks :)

ZolotukhinM commented:
Oh, please don't forget to update the commit message before landing too!

tugsbayasgalan committed:
Update on "Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement optimization bisect"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement optimization bisect"


Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement optimization bisect"


Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement optimization bisect"


Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
👍

tugsbayasgalan committed:
Update on "Implement optimization bisect"


Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implement optimization bisect"


Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

[ghstack-poisoned]
